File: spike/correlation-fastapi/src/sast.py
"""
Stand-in for the Semgrep adapter, Python side.

Emits the SAME FIELDS a Semgrep finding would — rule_id, cwe, file, line, and
an enclosing symbol resolved from the AST rather than a bare line number —
because those are what the join and the fingerprint consume.

RULE kase.idor.missing-ownership-predicate (CWE-639)

Fires when an ORM read inside a method that receives a requester/owner
identifier filters on the record id alone, never constraining on the caller.
Matches the idiomatic SQLAlchemy shape:

    session.query(Invoice).filter_by(id=invoice_id).first()
"""

import ast
import hashlib
import json
import os
from dataclasses import dataclass, field
from typing import Any, Optional

from .codemap import SourceLocation

RULE_ID = "kase.idor.missing-ownership-predicate"
CWE = "CWE-639"

OWNER_PARAM_HINTS = {"requester_id", "user_id", "owner_id", "current_user_id", "actor_id"}
OWNERSHIP_FIELD_HINTS = {"owner_id", "user_id", "account_id", "tenant_id", "organization_id"}
FILTER_CALLS = {"filter_by", "find_unique", "find_first", "get_by"}


@dataclass
class Evidence:
    id: str
    tool: str
    type: str
    evidence_class: str          # 'replayable' | 'observational'
    sha256: str
    payload: Any


@dataclass
class Finding:
    id: str
    origin: str
    tool: str
    rule_id: Optional[str]
    category: str
    cwe: Optional[str]
    title: str
    severity_proposed: str
    affected_target: Optional[str]
    source_locations: list[SourceLocation] = field(default_factory=list)
    evidence: list[Evidence] = field(default_factory=list)
    description: str = ""


def sha256(s: str) -> str:
    return hashlib.sha256(s.encode("utf-8")).hexdigest()


def short_hash(s: str) -> str:
    return sha256(s)[:8]


def run_sast(root: str) -> list[Finding]:
    root = os.path.abspath(root)
    findings: list[Finding] = []

    for dirpath, dirnames, filenames in os.walk(root):
        dirnames[:] = [d for d in dirnames if d not in {"__pycache__", ".venv", "node_modules"}]
        for fn in filenames:
            if not fn.endswith(".py"):
                continue
            full = os.path.join(dirpath, fn)
            rel = os.path.relpath(full, root).replace(os.sep, "/")
            try:
                tree = ast.parse(open(full, encoding="utf-8").read())
            except (SyntaxError, UnicodeDecodeError):
                continue

            for cls in [n for n in tree.body if isinstance(n, ast.ClassDef)]:
                for method in [
                    n for n in cls.body if isinstance(n, (ast.FunctionDef, ast.AsyncFunctionDef))
                ]:
                    owner_param = _owner_param(method)
                    if not owner_param:
                        continue

                    for node in ast.walk(method):
                        if not isinstance(node, ast.Call):
                            continue
                        func = node.func
                        if not isinstance(func, ast.Attribute):
                            continue
                        if func.attr not in FILTER_CALLS:
                            continue

                        fields = _filter_fields(node)
                        if fields is None:
                            continue
                        if any(f in OWNERSHIP_FIELD_HINTS for f in fields):
                            continue

                        symbol = f"{cls.name}.{method.name}"
                        line = node.lineno
                        excerpt = ast.unparse(node)

                        findings.append(
                            Finding(
                                id=f"WB-{short_hash(f'{rel}:{symbol}:{RULE_ID}')}",
                                origin="whitebox",
                                tool="semgrep-equivalent",
                                rule_id=RULE_ID,
                                category="authorization",
                                cwe=CWE,
                                title="Resource lookup lacks an ownership predicate",
                                severity_proposed="critical",
                                affected_target=None,
                                source_locations=[
                                    SourceLocation(file=rel, enclosing_symbol=symbol, line=line)
                                ],
                                evidence=[
                                    _evidence(rel, symbol, line, excerpt, owner_param, fields)
                                ],
                                description=(
                                    f"`{symbol}` accepts `{owner_param}` but filters on "
                                    f"{{ {', '.join(fields)} }} only. The caller identity never "
                                    f"reaches the query predicate, so any authenticated caller "
                                    f"can read any record by id."
                                ),
                            )
                        )

    return findings


def _owner_param(fn: ast.FunctionDef) -> Optional[str]:
    for arg in [*fn.args.posonlyargs, *fn.args.args, *fn.args.kwonlyargs]:
        if arg.arg in OWNER_PARAM_HINTS:
            return arg.arg
    return None


def _filter_fields(call: ast.Call) -> Optional[list[str]]:
    """Returns the keyword names passed to the filter call, or None if none."""
    names = [kw.arg for kw in call.keywords if kw.arg]
    return names or None


def _evidence(
    file: str, symbol: str, line: int, excerpt: str, owner_param: str, fields: list[str]
) -> Evidence:
    payload = {
        "ruleId": RULE_ID,
        "cwe": CWE,
        "file": file,
        "enclosingSymbol": symbol,
        "line": line,
        "excerpt": excerpt,
        "ownerParamAvailable": owner_param,
        "queryConstrainedOn": fields,
    }
    blob = json.dumps(payload, sort_keys=True)
    return Evidence(
        id=f"EV-{short_hash(blob)}",
        tool="semgrep-equivalent",
        type="sast_json",
        # Rule + file + symbol is re-derivable from source: replayable.
        evidence_class="replayable",
        sha256=sha256(blob),
        payload=payload,
    )

File: spike/correlation/src/sast.ts
import { Project, SyntaxKind, MethodDeclaration } from 'ts-morph';
import * as crypto from 'crypto';
import * as path from 'path';
import { Finding, Evidence } from './types';

/**
 * Stand-in for the Semgrep adapter.
 *
 * The real system runs Semgrep (docs/07 §4.8). Semgrep is a Python tool and
 * is not installed here, so this implements one equivalent rule directly on
 * the AST. It deliberately emits the SAME FIELDS a Semgrep finding would —
 * ruleId, cwe, file, line, and an enclosing symbol resolved from the AST
 * rather than a bare line number — because those fields are what the join
 * and the fingerprint consume.
 *
 * What this does NOT prove: Semgrep's rule coverage or accuracy. That was
 * never the uncertain part. What it does prove is that a SAST-shaped finding
 * carrying an enclosing symbol joins cleanly to a route-derived source
 * location, which is the thing the spike exists to test.
 *
 * RULE kase.idor.missing-ownership-predicate (CWE-639)
 *
 * Fires when a Prisma read inside a method that receives a requester/owner
 * identifier filters by record id alone, never constraining on the caller.
 */
const RULE_ID = 'kase.idor.missing-ownership-predicate';
const CWE = 'CWE-639';

const OWNER_PARAM_HINTS = ['requesterid', 'userid', 'ownerid', 'currentuserid', 'actorid'];
const OWNERSHIP_FIELD_HINTS = ['ownerid', 'userid', 'accountid', 'tenantid', 'organizationid'];
const PRISMA_READS = ['findUnique', 'findFirst', 'findMany'];

export function runSast(tsConfigFilePath: string): Finding[] {
  const project = new Project({ tsConfigFilePath });
  const root = path.dirname(tsConfigFilePath);
  const findings: Finding[] = [];

  for (const sf of project.getSourceFiles()) {
    const relFile = path.relative(root, sf.getFilePath()).split(path.sep).join('/');
    if (relFile.includes('node_modules')) continue;

    for (const cls of sf.getClasses()) {
      for (const method of cls.getMethods()) {
        const ownerParam = findOwnerParam(method);
        if (!ownerParam) continue;

        for (const call of method.getDescendantsOfKind(SyntaxKind.CallExpression)) {
          const callee = call.getExpression();
          if (!callee.isKind(SyntaxKind.PropertyAccessExpression)) continue;
          if (!PRISMA_READS.includes(callee.getName())) continue;

          const whereFields = extractWhereFields(call);
          if (whereFields === null) continue;

          const constrainsOwnership = whereFields.some((f) =>
            OWNERSHIP_FIELD_HINTS.includes(f.toLowerCase()),
          );
          if (constrainsOwnership) continue;

          const enclosingSymbol = `${cls.getName()}.${method.getName()}`;
          const line = call.getStartLineNumber();
          const excerpt = call.getText();

          findings.push({
            id: `WB-${shortHash(`${relFile}:${enclosingSymbol}:${RULE_ID}`)}`,
            origin: 'whitebox',
            tool: 'semgrep-equivalent',
            ruleId: RULE_ID,
            category: 'authorization',
            cwe: CWE,
            title: 'Resource lookup lacks an ownership predicate',
            severityProposed: 'critical',
            affectedTarget: null,
            sourceLocations: [{ file: relFile, enclosingSymbol, line }],
            evidence: [sastEvidence(relFile, enclosingSymbol, line, excerpt, ownerParam, whereFields)],
            description:
              `\`${enclosingSymbol}\` accepts \`${ownerParam}\` but queries on ` +
              `{ ${whereFields.join(', ')} } only. The caller identity never reaches the ` +
              `query predicate, so any authenticated caller can read any record by id.`,
          });
        }
      }
    }
  }

  return findings;
}

function findOwnerParam(method: MethodDeclaration): string | null {
  for (const p of method.getParameters()) {
    if (OWNER_PARAM_HINTS.includes(p.getName().toLowerCase())) return p.getName();
  }
  return null;
}

/** Returns the property names inside `{ where: { ... } }`, or null if absent. */
function extractWhereFields(call: import('ts-morph').CallExpression): string[] | null {
  const arg = call.getArguments()[0];
  if (!arg || !arg.isKind(SyntaxKind.ObjectLiteralExpression)) return null;

  const whereProp = arg.getProperty('where');
  if (!whereProp || !whereProp.isKind(SyntaxKind.PropertyAssignment)) return null;

  const initializer = whereProp.getInitializer();
  if (!initializer || !initializer.isKind(SyntaxKind.ObjectLiteralExpression)) return null;

  return initializer
    .getProperties()
    .map((p) => {
      if (p.isKind(SyntaxKind.PropertyAssignment)) return p.getName();
      if (p.isKind(SyntaxKind.ShorthandPropertyAssignment)) return p.getName();
      return null;
    })
    .filter((n): n is string => n !== null);
}

function sastEvidence(
  file: string,
  symbol: string,
  line: number,
  excerpt: string,
  ownerParam: string,
  whereFields: string[],
): Evidence {
  const payload = {
    ruleId: RULE_ID,
    cwe: CWE,
    file,
    enclosingSymbol: symbol,
    line,
    excerpt,
    ownerParamAvailable: ownerParam,
    queryConstrainedOn: whereFields,
  };
  return {
    id: `EV-${shortHash(JSON.stringify(payload))}`,
    tool: 'semgrep-equivalent',
    type: 'sast_json',
    // Rule + file + symbol is re-derivable from source: replayable, gate-eligible.
    evidenceClass: 'replayable',
    sha256: sha256(JSON.stringify(payload)),
    payload,
  };
}

export function sha256(input: string): string {
  return crypto.createHash('sha256').update(input).digest('hex');
}

export function shortHash(input: string): string {
  return sha256(input).slice(0, 8);
}
